src/irmin-pack/inode.ml

```diff
@@ -759,6 +759,117 @@ struct
       let hash t = I.hash t.v
       let stable t = I.stable t.v
       let length t = I.length t.v
+
+      module Serde = struct
+        type sexp = Sexplib.Sexp.t
+
+        (* type ptr = { target_hash : hash Lazy.t; mutable target : t option }
+         * and tree = { depth : int; length : int; entries : ptr option array }
+         * and v = Values of value StepMap.t | Tree of tree
+         * and t = { hash : hash Lazy.t; stable : bool; v : v } *)
+
+        let parse_from_string s =
+          let lexbuf = Lexing.from_string s in
+          Sexplib.Sexp.scan_sexp lexbuf
+
+        let parse_from_file f = Sexplib.Sexp.load_sexp f
+        let sexp_of_hash h = Sexplib.Sexp.(Atom (Fmt.str "%a" pp_hash h))
+        let sexp_of_lazy_hash h = Sexplib.Std.sexp_of_lazy_t sexp_of_hash h
+
+        let sexp_of_metadata m =
+          Sexplib.Sexp.Atom (Fmt.str "%a" (Irmin.Type.pp metadata_t) m)
+
+        (* A kind is either a Content (B for blob) or a Node (N) *)
+        let sexp_of_kind k =
+          let open Sexplib.Sexp in
+          match k with
+          | `Contents (h, m) ->
+              List [ Atom "B"; sexp_of_hash h; sexp_of_metadata m ]
+          | `Node h -> List [ Atom "N"; sexp_of_hash h ]
+
+        let sexp_of_step s =
+          Sexplib.Sexp.Atom (Fmt.str "%a" (Irmin.Type.pp step_t) s)
+
+        let rec list_of_entries a =
+          Array.fold_left
+            (fun acc e ->
+              match e with
+              | None -> acc
+              | Some p -> (
+                  match p.I.target with
+                  | None -> acc
+                  | Some t -> List.rev_append (list_of_t t) acc))
+            [] a
+
+        and list_of_tree t = list_of_entries t.I.entries
+
+        and list_of_stepmap vs =
+          let open Sexplib.Sexp in
+          StepMap.fold
+            (fun s k acc -> List [ sexp_of_step s; sexp_of_kind k ] :: acc)
+            vs []
+
+        and list_of_v v =
+          match v with
+          | I.Values vs -> list_of_stepmap vs
+          | Tree t -> list_of_tree t
+
+        and list_of_t t = list_of_v t.I.v
```

Comment:
I think you could have used `I.list` to retrieve the `(step * value) list` of `t`, instead of iterating yourself